# PT-AML

## Description
This repository contains all the data, code, model and results for the PT-AML project

We get input from the BeatAML project where Wave 1+2 form the training set and Wave 3+4 form the test set.
The input includes clinical, DNAseq, RNAseq, cell state, module information for each patient.
For each patient, we have information about several drugs used for treatment including the IC50 score.


## Objectives
1. Build a model to learn the intricate relationship between omics, disease state and drug response.
2. Predict the right drug for a given patient with multi-omics profile + disease state information.
3. Given a drug (new) find the right cohort of patients for which the drug will induce best response.


## Goals
1. Extract clinical, transcriptomic, dnaseq, celltype, module features common to Wave 1+2 (train) and Wave 3+4 (test).
2. Compare the multi-omics features between train and test through 2d-embedding using t-SNE to show uniformity of representations.
3. Learn feature representations for drugs using molecular fingerprint or a SMILES encoder (from drugrepurpose paper i.e. Mall et al).
4. Estimate hallmark pathway enrichment relevant for cancer using GSVA technique.
5. Estimate sample specific distance of pathways from a drug using String PPI, random walk approach and sample transcriptomic data together in a ranking setting (AUC).
6. Perform comprehensive comparison of machine learning models on features of drug and samples combined.
7. Perform explanability analysis using SHAPley method.

## Deadlines
- Project Proposal: [Jun 2023]
- Code Check in: [August 2023]
- Manuscript Draft: [September 2024]

## Pre-processing scripts
1. `analyze_cell_lines_rnaseq.R`: Loads gene expression and clinical information from Wave 1+2 and Wave 3+4, perform t-SNE on gene expression data, merge gene expression, clinical annotations and t-SNE co-ordinates. Add pathway enrichments, cell type enrichments and module enrichments for each sample. Finally, oncogenes and 150 most variable genes are identified and only expression of those genes are considered for gene expression profile. The training and test data are saved as Traiing_Set_with_Onco_Var_Expr_Clin_PA_CTS.csv and Test_Set_with_Onco_Var_Expr_Clin_PA_CTS.csv. The metadata is saved in TRAINING_METADATA.csv file in the Results folder. Figures generated using this script include Supp. Fig 1A.

2. `gene_expression_data_analysis.R`: Performs variance analysis of gene expression data and identifies the optimal set of genes to select which divide the data into optimal number of clusters. Figures generated using this script include Supp. Fig 1B,C and D.

3. `analyze_cell_lines_dnaseq.R`: Gets the dnaseq data and divides into training and test set and converts the dnaseq data into sample by mutation type matrix for training and test sets.

4. `preprocess_drugs.py`: Get list of drug names from BeatAML dataset and corresponidng compound id from pubchempy package in python and get properties including InChiKey, SMILES, XLogP, Molecular Weight. Run the SMILES based encoding representation by calling 'ls_generator2.py' and save it in Drug_Full_SMILES_Embedding.csv. Figure 3A is representation of the embedding vector generation process.

5. `morgan_fps.py`: Get molecular fingerprints for all drugs from BeatAML dataset

6. `analyze_cell_lines_drug_combinations.R`: Takes input sample info, drug info, train + test feature files. Based on drug targets and string ppi network estimates propagation/diffusion score for each drug. Combines the diffusion score with the gene expression profile in the sample and then estimates distance of each pathway (geneset) from the drug using AUCell package. Final output is Revised_Train_Set_with_IC50.csv and Revised_Test_Set_with_IC50.csv. Figure 3B is representative of the random walk with restart based diffusion score combined with expression profile of sample. This is followed by AUCell for each pathway to estimate the distance of the pathway from a given drug.

7. `preprocess_cellline.py`: Cleans the sample features including clinical, gene expression, mutation matrix and generates training and test pickle sets.

8. `make_fig1b.R`: Make a complexheatmap of the oncogenes + top varying genes + clinical characteristics (from BeatAML) + pathway enrichments + celltype enrichments + module enrichments + mutation type matrix. Figure 1 is generated through this script.

9. `combined_cell_drug_info.py`: Runs the `preprocess_cellline.py` file first and combines with Revised_Train_Set_with_IC50.csv and Revised_Test_Set_with_IC50.csv. Drug representations generated by SMILES encoder or MFP are used to generate two training sets - a) Training_Set_Var_with_Drug_Embedding_Cell_Info.pkl b) Training_Set_Var_with_Drug_MFP_Cell_Info.pkl and corresponding test sets. Figure 4A is generated by this script.

10. `all_cells_functions.R`: Miscallenous R script to plot figures and additional functions.

## ML scripts  

1. `glr_model.py`: Builds a generalized linear model with 5-fold cross-validation and hyper-parameter search and predictions on test set. Figures 4B and Supp. Figure 2A are generated by this code.

2. `svr_model.py`: Builds support vector regressors model with 5-fold cross-validation and hyper-parameter search and predictions on test set. Figures 4C and Supp. Figure 2B are generated by this code.

3. `rf_model.py`: Builds random forest model with 5-fold cross-validation and hyper-parameter search and predictions on test set. Figures 4D and Supp. Figure 2C are generated by this code. Supp. Figure 3B generated by this code highlights top random forest features for predictions.

4. `xgb_model.py`: Builds XGBoost model with 5-fold cross-validation and hyper-parameter search and predictions on test set. Figures 4E and Supp. Figure 2D are generated by this code.
 
5. `lightgbm_model.py`: Builds LightGBM model with 5-fold cross-validation and hyper-parameter search and predictions on test set. Figures 4F and Supp. Figure 2E are generated by this code. Supp. Figure 3C generated by this code highlights top lightgbm featuers for predictions.

6. `catboost_model.py`: Builds CatBoost model with 5-fold cross-validation and hyper-parameter search and predictions on test set. Figures 4G and Supp. Figure 2F are generated by this code. Supp. Figure 3A generated by this code highlights top catboost features for predictions.

7. `nn_model.py`: Builds neural network model with 5-fold cross-validation and hyper-parameter search and predictions on test set. Figures 4H and Supp. Figure 2G are generated by this code.

8. `catboost_ablation_model.py`: Ablation study with the best model (catboost with MFP for drugs) done in 5-fold cross-validation settings. Supp. Figure 2H-S are generated by this code. Supp. Table 1 generated by this code.

9. `gs_cv_info.py`: Cross-validation and test results for machine learning models as highlighted in Table 1.

## DL scripts

1. `cnn_model.py`: Builds CNN based regression model starting with SMILES representation for drugs and feature representation of samples used in FFNN. Figure 4J generated by this code.

2. `lstm_model.py`: Builds LSTM based regression model starting with SMILES representation for drugs and feature representation of samples used in FFNN. Figure 4K generated by this code.

3. `gat_model.py`: Builds Graph Attenion based regression model starting with SMILES representation for drugs and feature representation of samples used in FFNN. Figure 4I generated by this code. Figure 3C is graphical representation of GAT model.

## Results

1. `make_heatmap_sample_drug.R`: Takes as input predictions for all drug and patient combinations from the best model (catboost with MFP drug features) and plots the same as a heatmap. Figure 5 is generated by this code.

2. `correlation_comparison.R`: Takes as input predictions for all drug and patient combinations from the best model (Catboost with MFP drug features). Figure 4L and 4M are generated by this code.
